Tests/RedisTests/RedisDataDecoderTests.swift

```diff
@@ -0,0 +1,324 @@
+        try errorTestCase(protocolString: "-ERROR\r\n-OTHER ERROR\r\n",
+                          expectedString: "ERROR",
+                          otherString: "OTHER ERROR")
+    }
+
+    private func integerTestCase(protocolString: String,
+                                      expectedInt: Int?,
+                                      otherInt: Int? = nil) throws {
+        let embeddedChannel = EmbeddedChannel()
+        try embeddedChannel.pipeline.add(handler: decoder).wait()
+        var buff = allocator.buffer(capacity: 256)
+        buff.write(string:  protocolString)
+        try embeddedChannel.writeInbound(buff)
+        let decoded: RedisData? = embeddedChannel.readInbound()
+        let otherInbound: RedisData? = embeddedChannel.readInbound()
+
+        XCTAssertEqual(decoded?.int, expectedInt)
+        XCTAssertEqual(otherInbound?.int, otherInt)
+        _ = try embeddedChannel.finish()
+    }
+
+    func testIntegers() throws {
+        try assertWillParseNil(string: ":100")
+        try assertWillParseNil(string: ":100\r")
+        try integerTestCase(protocolString: ":1000\r\n",
+                            expectedInt: 1000)
+
+        try integerTestCase(protocolString: ":1000\r\n:99\r\n",
+                            expectedInt: 1000,
+                            otherInt: 99)
+    }
+
+    private func bulkStringTestCase(protocolString: String,
+                                      expectedString: String?,
+                                      otherString: String? = nil) throws {
+        let embeddedChannel = EmbeddedChannel()
+        try embeddedChannel.pipeline.add(handler: decoder).wait()
+        var buff = allocator.buffer(capacity: 256)
+        buff.write(string:  protocolString)
+        try embeddedChannel.writeInbound(buff)
+        let decoded: RedisData? = embeddedChannel.readInbound()
+        let otherInbound: RedisData? = embeddedChannel.readInbound()
+
+        XCTAssertEqual(decoded?.string, expectedString)
+        XCTAssertEqual(otherInbound?.string, otherString)
+        _ = try embeddedChannel.finish()
+    }
+
+    func testBulkString() throws {
+        try assertWillParseNil(string: "$0\r\n\r")
+        try bulkStringTestCase(protocolString: "$0\r\n\r\n",
+                               expectedString: "")
+        try assertWillParseNil(string: "$1\r\na\r")
+        try bulkStringTestCase(protocolString: "$1\r\na\r\n", expectedString: "a")
+        try assertWillParseNil(string: "$3\r\naaa\r")
+        try bulkStringTestCase(protocolString: "$3\r\nfoo\r\n",
+                               expectedString: "foo")
+        try bulkStringTestCase(protocolString: "$3\r\naaa\r\n", expectedString: "aaa")
+        try bulkStringTestCase(protocolString: "$1\r\na\r\n$2\r\naa\r\n",
+                               expectedString: "a",
+                               otherString: "aa")
+
+        // decode special charachters guard against multibyte decoding
```

Comment:
nice tests :)

Reply:
I tried... the tests for implementing `AMQP` are going to be rediculous